### DISCUSSION TITLE
Can I use environment variables inside a manifest that is deployed via Argo?

### DISCUSSION THREAD

**Discussion by loganmarchione** [argoproj/argo-cd#10849] [Q&A] [answered]
I'm learning K8s, so bear with me as a noob.

I'm running a single-node K3s cluster at home, and have successfully deployed the [traefik/whoami][1] application using the command below, but would like to deploy it via ArgoCD.

    cat apps/whoami/whoami.yaml | envsubst | kubectl apply -f -

The manifest I created.

    ---
    apiVersion: v1
    kind: Namespace
    metadata:
      name: k3s-test
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: whoami-deploy
      namespace: k3s-test
      labels:
        app: whoami
    spec:
      replicas: 3
      selector:
        matchLabels:
          app: whoami
      template:
        metadata:
          labels:
            app: whoami
        spec:
          containers:
            - name: whoami
              image: traefik/whoami:v1.8.0
              ports:
                - containerPort: 80
    ---
    apiVersion: v1
    kind: Service
    metadata:
      name: whoami-svc
      namespace: k3s-test
      labels:
        service: whoami
    spec:
      type: ClusterIP
      ports:
        - name: http
          port: 80
          protocol: TCP
      selector:
        app: whoami
    ---
    apiVersion: networking.k8s.io/v1
    kind: Ingress
    metadata:
      name: whoami-ingress
      namespace: k3s-test
      annotations:
        traefik.ingress.kubernetes.io/router.entrypoints: web
    spec:
      rules:
        - host: whoami.${DOMAIN_NAME}
          http:
            paths:
              - path: /
                pathType: Prefix
                backend:
                  service:
                    name: whoami-svc
                    port:
                      number: 80

I want to publish my code to GitHub so that ArgoCD can sync it, but don't want to expose information that is not necessarily secret, but not necessarily public. Currently, my domain name is set as an environment variable (because I don't want to commit mydomain.com to my GitHub repo) and I'm using `envsubst` when running `kubectl apply`. Does ArgoCD have similar functionality? I found [this GitHub issue][2] showing ArgoCD probably doesn't have variable interpolation, but is there an alternative? Or do I need to store my domain name as a full-on K8s secret?

  [1]: https://hub.docker.com/r/traefik/whoami
  [2]: https://github.com/argoproj/argo-cd/issues/5580

**Comment by jgwest** [comment, marked as answer]
The available customizations vary based on which plugin/tool you are using to generate your manifest.

A couple options that come to mind, but these are more advanced if you are new to K8s:
- `ApplicationSet` can be used to create Applications using data from [ApplicationSet resource or external sources](https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/).
- If you are willing to use Helm, you can [Helm Parameters](https://argo-cd.readthedocs.io/en/stable/user-guide/helm/#helm-parameters) 

I'm not aware of a way to provide custom values to [kustomize](https://argo-cd.readthedocs.io/en/stable/user-guide/kustomize/) beyond those described in the docs, or to flat YAML files (which is what I presume you have in your Git repository).
